designs/sam-config.md

```diff
@@ -0,0 +1,369 @@
+parameter_overrides="ParameterKey=KeyPairName,ParameterValue=MyKey ParameterKey=InstanceType,ParameterValue=t1.micro"
+```
+
+Eventual merged configuration read during `sam build` in-memory.
+
+```
+[default.build.parameters]
+use_container = True
+skip_pull_image = True
+parameter_overrides="ParameterKey=KeyPairName,ParameterValue=MyKey ParameterKey=InstanceType,ParameterValue=t1.micro"
+```
+
+Open Questions
+-------------------------------
+
+* Potentially every sam command could have functionality to have a series of command line parameters exported into a configuraion file.
+
+
+Out-of-Scope
+------------
+
+* Not focusing on a global configuration. SAM CLI already has a notion of a global config at `~/.aws-sam/metadata.json`
+
+User Experience Walkthrough
+---------------------------
+
+Once a configuration file is appropriately populated, day to day workflows per application developed with SAM CLI become much simpler.
+
+* `sam build` -> `sam package` -> `sam deploy`
+* `sam build` -> `sam local invoke`
+* `sam build` -> `sam package` -> `sam publish`
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+New command line argument is added per command called `--env` to be able to specify non default environment section within a config file.
+
+
+### Breaking Change
+
+* No breaking changes to the CLI, in absence of the configuration file, CLI continues to work as normal.
+
+Design
+------
+
+*Explain how this feature will be implemented. Highlight the components
+of your implementation, relationships* *between components, constraints,
+etc.*
+
+A custom decorator to `click.option` is developed which reads from a configuration file the sections that are pertinent to that particular command and populates the click's `map` context.
+
+The configuration file parser is a custom provider that can be made to understand any configuration file format in a pluggable manner.
+
+This decorator benefits from the same type checking that some SAM CLI parameters already use.
+
+A custom callback function (`configuration_callback`) (for the click option) that takes in a custom configuration parser (`provider`) will have rules in place, on how the corresponding configuration can be retrieved and what are the parts that the configuration parser has access to read from.
+
+```
+provider = attrs.pop("provider", TomlProvider(rules=DefaultRules(), command="build", section="parameters"))
+attrs["type"] = click.STRING
+saved_callback = attrs.pop("callback", None)
+partial_callback = functools.partial(onfiguration_callback, cmd_name, option_name, env_name, saved_callback, provider)
+attrs["callback"] = partial_callback
+click.option(*param_decls, **attrs)(f)
+
+```
+
+Phases
+------
+
+The design can be built in phases.
+
+* No option to specify configuration file or env name
+* Specify configuration file with an environment variable
+* Read `--env` to make sure we can select an appropriate portion in configuration file.
+
+
+`.samrc` Changes
```

Comment:
This PR is `samrc`. Can you actually rename this part of the design doc template to call it `samconfig` instead?

Reply:
sure.